### PR TITLE
feat: expose internal state of composite price through the API

### LINKED ISSUE #10502: [API]: Provide current "vega side price for funding" on an API endpoint
Currently the `.../api/v2/markets/data` endpoint provides:
- mark price, 

and for perps also
```
"productData": {
                "perpetualData": {
                    "fundingPayment": "-17994",
                    "fundingRate": "-0.001210715416813",
                    "internalTwap": "14844293",
                    "externalTwap": "14862287",
                    "seqNum": "160",
                    "startTime": "1706544889000000000",
                    "indexPrice": "1494299",
                    "nextIndexPriceCalc": "1706556424211855752",
                    "indexPriceType": "COMPOSITE_PRICE_TYPE_LAST_TRADE"
                }
            },
``` 

Request (minimal): add the "vega side price for funding" next to the "indexPrice". 

Bonus (if possible): for both (perps markets) and mark price only (futures) of the composite price show the values of the 
1)~market~ ~config~ ~used~ ~somewhere~ ~(method,~weights,~staleness~settings)~ -- apparently available elsewhere so not needed here
2) values of the component prices going in
3) their age

### PR DESCRIPTION
**Pull request by ze97286**
Closes #10502

### CODE REVIEW

File: datanode/gateway/graphql/schema.graphql [hunk 1]
```diff
@@ -245,6 +245,24 @@ type PerpetualData {
   internalCompositePriceType: CompositePriceType!
   "RFC3339Nano time indicating the next time internal composite price will be calculated for perpetual markets, where applicable"
   nextInternalCompositePriceCalc: String!
+  "The last value from the external oracle"
+  underlyingIndexPrice: String!
+  "The internal state of the underlying internal composite price"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:250]:
```suggestion
  "Configuration for the internal composite price method"
```

File: datanode/gateway/graphql/schema.graphql [hunk 2]
```diff
@@ -400,6 +420,8 @@ type ObservableMarketData {
   productData: ProductData
   "The methodology used to calculated mark price"
   markPriceType: CompositePriceType!
+  "The internal state of the mark price composite"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:423]:
```suggestion
  "Configuration for the internal composite price method"
```

File: protos/sources/vega/vega.proto [hunk 3]
```diff
@@ -1329,6 +1333,18 @@ message MarketData {
...
   // The method used for calculating the mark price.
   CompositePriceType mark_price_type = 34;
+  // The internal state of the composite price.
+  CompositePriceState mark_price_state = 35;
+}
+
+message CompositePriceSource {
+	string price_source = 1;
+	string price = 2;
+  int64 last_updated = 3;
```
Comment (candida-d) [inline protos/sources/vega/vega.proto:1343]:
```suggestion
  int64 last_updated = 3;
  // Timestamp in Unix nanoseconds when the price source was last updated.
```

File: protos/sources/vega/vega.proto [hunk 4]
```diff
@@ -1329,6 +1333,18 @@ message MarketData {
   int64 next_network_closeout = 33;
   // The method used for calculating the mark price.
   CompositePriceType mark_price_type = 34;
+  // The internal state of the composite price.
+  CompositePriceState mark_price_state = 35;
+}
+
+message CompositePriceSource {
+	string price_source = 1;
```
Comment (candida-d) [inline protos/sources/vega/vega.proto:1341]:
```suggestion
	string price_source = 1;
	// Source of the price.
```

File: protos/sources/vega/vega.proto [hunk 5]
```diff
@@ -1329,6 +1333,18 @@ message MarketData {
...
+	string price = 2;
```
Comment (candida-d) [inline protos/sources/vega/vega.proto:1342]:
```suggestion
	string price = 2;
	// Current value of the composite source price.
```

File: protos/sources/vega/vega.proto [hunk 6]
```diff
@@ -1242,6 +1242,10 @@ message PerpetualData {
   int64 next_internal_composite_price_calc = 8;
   // The method used for calculating the internal composite price, for perpetual markets only.
   CompositePriceType internal_composite_price_type = 9;
+  // The last seen value of the settlement oracle.
+  string underlying_index_price = 10;
+   // The internal state of the composite price.
+  CompositePriceState internal_composite_price_state = 11;
```
Comment (candida-d) [inline protos/sources/vega/vega.proto:1248]:
```suggestion
  CompositePriceState internal_composite_price_state = 11;
  // Configuration for the internal composite price method.
```

File: protos/sources/vega/vega.proto [hunk 7]
```diff
@@ -1329,6 +1333,18 @@ message MarketData {
   int64 next_network_closeout = 33;
   // The method used for calculating the mark price.
   CompositePriceType mark_price_type = 34;
+  // The internal state of the composite price.
+  CompositePriceState mark_price_state = 35;
```
Comment (candida-d) [inline protos/sources/vega/vega.proto:1337]:
```suggestion
  CompositePriceState mark_price_state = 35;
  // "Configuration for the internal mark price method"
```

Comment (candida-d) [reply protos/sources/vega/vega.proto:1337]: @ze97286 Is the string I added correct?

File: protos/sources/vega/vega.proto [hunk 8]
```diff
@@ -1329,6 +1333,18 @@ message MarketData {
...
+}
+
+message CompositePriceSource {
+	string price_source = 1;
+	string price = 2;
+  int64 last_updated = 3;
+}
+
+message CompositePriceState {
```
Comment (candida-d) [inline protos/sources/vega/vega.proto:1346]:
```suggestion
message CompositePriceState {
// Underlying state of the composite price.
```

File: protos/sources/vega/vega.proto [hunk 9]
```diff
@@ -1329,6 +1333,18 @@ message MarketData {
   int64 next_network_closeout = 33;
   // The method used for calculating the mark price.
   CompositePriceType mark_price_type = 34;
+  // The internal state of the composite price.
+  CompositePriceState mark_price_state = 35;
+}
+
+message CompositePriceSource {
```
Comment (candida-d) [inline protos/sources/vega/vega.proto:1340]: Could you put a doc-string on this one?

File: protos/sources/vega/vega.proto [hunk 10]
```diff
@@ -1242,6 +1242,10 @@ message PerpetualData {
   int64 next_internal_composite_price_calc = 8;
   // The method used for calculating the internal composite price, for perpetual markets only.
   CompositePriceType internal_composite_price_type = 9;
+  // The last seen value of the settlement oracle.
+  string underlying_index_price = 10;
```
Comment (wwestgarth) [inline protos/sources/vega/vega.proto:1246]: To keep the terminology consistent this field should just be called `external_price`. Throughout all over the perpsy API we just use the terms internal/external.

File: core/execution/future/market.go [hunk 11]
```diff
@@ -4374,6 +4381,9 @@ func (m *Market) settlementDataPerp(ctx context.Context, settlementData *num.Num
 	if sdi == nil {
 		return
 	}
+
+	m.underlyingIndexPrice = sdi
```
Comment (wwestgarth) [inline core/execution/future/market.go:4385]: Here `sdi` is the funding-payment at the end of the period, and not the last external price received from the data-sources. If you want the last external price, you just need to grab:
```
externalPrice = p.externalTWAP.points[-1]
```
somewhere here:
https://github.com/vegaprotocol/vega/blob/develop/core/products/perpetual.go#L751

File: protos/sources/vega/vega.proto (hunk 10 above)
Comment (ze97286) [reply protos/sources/vega/vega.proto:1246]: I don't have a preference but this name is what the ticket has https://github.com/vegaprotocol/vega/issues/10502#issuecomment-1916645422

File: protos/sources/vega/vega.proto [hunk 12]
```diff
@@ -1242,6 +1242,10 @@ message PerpetualData {
...
+   // The internal state of the composite price.
```
[the remark this answers is not in the document]
Comment (ze97286) [reply protos/sources/vega/vega.proto:1247]: ` // Configuration for the internal composite price method.`

This is completely wrong @candida-d, what did you mean?

[the remark this answers is not in the document]
Comment (candida-d) [reply protos/sources/vega/vega.proto:1247]: @ze97286 changed them all to state

**Pull request merged**